File: python/recursion/sqrt.py
import math

def helper(n, fromIndex, toIndex):
    
    midpoint = (fromIndex + toIndex) / 2
    midpoint_squared = midpoint * midpoint

    # case midpoint_squared == sqrt(n)
    if (midpoint_squared == n or (abs(midpoint_squared - n) < 0.00001)):
        return midpoint

    # recurse from (midpoint:toIndex)   
    elif (midpoint_squared < n):
        return helper(n, midpoint, toIndex)

    # recurse from (fromIndex:midpoint)
    else:
        return helper(n, fromIndex, midpoint)


def findSquareRoot(n):
    # ensure n is valid
    if not n or n < 0:
        print('invalid input')
        return

    i = 1
    found = False
    while not found:
        # case perfect square found
        if i * i == n:
            found = True
            return i
        elif i * i > n:
            # recursion case
            result = helper(n, i - 1, i)
            found = True
            return result
        i += 1


def main():
    n = 11
    result = findSquareRoot(n)
    print("{0:.5f}".format(result))

main()

File: js/recursion/sqrt.js
/*
 */

function sqrt(n) {
  let i = 1;
  let found = false;
  while (!found) {
    // base case
    if (i * i === n) {
      found = true;
      return i;
    } else if (i * i > n) {
      // recurese
      found = true;
      return helper(n, i - 1, i);
    }
    // increment i every time
    i += 1;
  }
}

function helper(n, fromIndex, toIndex) {
  const midpoint = (fromIndex + toIndex) / 2;
  const midpointSquared = Math.pow(midpoint, 2);

  // case midpointSquared === sqrt(n)
  if (midpointSquared === n || Math.abs(midpointSquared - n) < 0.00001) {
    return midpoint;
  } else if (midpointSquared < n) {
    // recurse only from [midpointSquared: toIndex]
    return helper(n, midpoint, toIndex);
  } else {
    // recurse only from [fromIndex: midpointSquared]
    return helper(n, fromIndex, midpoint);
  }
}

console.log(sqrt(24));
